### DISCUSSION TITLE
Seed and BatchRunner

### DISCUSSION THREAD

**Discussion by ed-ij** [mesa/mesa#1009] [Q&A] [answered]
It's a little unclear from the documentation so just hoping for some clarification based on the behavior I am observing.

From the fact that the reporters are collecting different data and my understanding of the code it looks like batch runner is varying the seed for every run of the model. The documentation isn't clear that this is the case but it is the behavior I would expect. 

1. Is this correct?

2. Is there any way to record the seeds being used for later analysis of specific runs of the model if necessary? I realize this could be done by varying and setting the seed externally but given this would seem to be duplicating what batch runner is already doing I would like to avoid doing that!

Thanks in advance!

**Comment by tpike3** [comment, marked as answer]
@edward-ij  Sorry for the delay. 

For 1. Yes, that is correct the seed varies for each run unless set. 

For 2: You should be able to collect the seend using the attribute from is initiation in the Model class by calling [Model.random](https://github.com/projectmesa/mesa/blob/bbf1b885ebe274e4b1eb269e30567a475e158aaf/mesa/model.py#L19)

**Reply by ed-ij** [reply]
Thanks, I'll test this and see if I can get it to work.

**Reply by ed-ij** [reply]
I see what you were saying. To be clear the behaviour appears to be:

- If you set the seed in your model initialisation you can collect it. 
- If you don't set the seed, the `model._seed` attribute will be set to `None`. As random.Random() does not return the generated seed alongside the generator it is not possible to identify the seed.

To get random behaviour but keep track of the seed you would have to pre-generate the seeds, for most consistent performance this would be best done when initialising the model:
```
class MyModel(Model):
    def __init__(self, var1, var2, seed):
        self.seed = seed
        if self.seed is None:
            self.seed = random.randrange(sys.maxsize)
        super().__init__(seed=self.seed)
        ...
```
This is a combination of a suggestion [(here)](https://stackoverflow.com/a/5012617) and comments made by @rht on issues related to determinism. This setup should enable random behaviour of the model as normal but also allow the seed to be recorded through data collection if desired.

**Reply by yinzhu5** [reply]
Hi, Thanks for your seeding explanation. I've implemented your suggestion, but I'm still seeing variable results even with the same seed. Should I also add random.seed(self.seed) after the super() call to ensure deterministic behavior? Are there other steps needed?

**Reply by rht** [reply]
That sounds like a bug. Needs a minimal reproducer.
